src/app/ReadHandler.cpp

```diff
@@ -82,6 +82,8 @@ CHIP_ERROR ReadHandler::OnReadRequest(Messaging::ExchangeContext * apExchangeCon
     if (err != CHIP_NO_ERROR)
     {
         ChipLogFunctError(err);
+        // Keep Shutdown() from double-closing our exchange.
+        mpExchangeCtx = nullptr;
```

Comment:
Where we close the mpExchangeCtx after ProcessReadRequest? I see we called mpExchangeCtx->WillSendMessage() within ProcessReadRequest to keep this exchang open

Reply:
`ProcessReadRequest` only calls `WillSendMessage` if it's returning success.  So there are two cases:

1. ProcessReadRequest returns success.  Then it called `WillSendMessage` and it's responsible for closing the exchange later.  In that case we have err == CHIP_NO_ERROR here and we do NOT null out `mpExchangeCtx`, so we can close it later.
2. ProcessReadRequest returns failure.  In that case it has not called `WillSendMessage`, the exchange will auto-close, we enter this block, and we need to keep `Shutdown` from doing an extra close on the exchange.

Reply:
so this logic require us not add any logic within ProcessReadRequest to return error after following code, right?

    if (mpExchangeCtx)
    {
        mpExchangeCtx->WillSendMessage();
    }

Reply:
That is correct.  I will add a comment to that effect.

Reply:
If ProcessReadRequest returns failure, I think the auto-close is reached after Shutdown,  right?

```
    err           = ProcessReadRequest(std::move(aPayload));

    if (err != CHIP_NO_ERROR)
    {
        ChipLogFunctError(err);
        // Keep Shutdown() from double-closing our exchange.
        mpExchangeCtx = nullptr;
        Shutdown();
    }
```
Do we set mpExchangeCtx = nullptr before or after the auto-close in this case?

Reply:
If `ProcessReadRequest` returns failure, then:

1. We never call `WillSendMessage`.
2. We null out `mpExchangeCtx` before the auto-close.
3. We call `Shutdown` before the auto-close, but that does not touch the exchange because `mpExchangeCtx` is null.
4. The auto-close happens when the stack unwinds, after `ReadHandler::OnReadRequest` returns.

Reply:
I see, the stack unwind will make sure the exchange is not leaked after we nullify the mpExchangeCtx